lib/std/static_string_map.zig

```diff
@@ -0,0 +1,502 @@
+const std = @import("std.zig");
+const mem = std.mem;
+
+/// Static string map optimized for small sets of disparate string keys.
+/// Works by separating the keys by length at initialization and only checking
+/// strings of equal length at runtime.
+pub fn StaticStringMap(comptime V: type) type {
+    return StaticStringMapWithEql(V, defaultEql);
+}
+
+/// Like `std.mem.eql`, but takes advantage of the fact that the lengths
+/// of `a` and `b` are known to be equal.
+pub fn defaultEql(a: []const u8, b: []const u8) bool {
+    if (a.ptr == b.ptr) return true;
+    for (a, b) |a_elem, b_elem| {
+        if (a_elem != b_elem) return false;
+    }
+    return true;
+}
+
+/// Like `std.ascii.eqlIgnoreCase` but takes advantage of the fact that
+/// the lengths of `a` and `b` are known to be equal.
+pub fn eqlAsciiIgnoreCase(a: []const u8, b: []const u8) bool {
+    if (a.ptr == b.ptr) return true;
+    for (a, b) |a_c, b_c| {
+        if (std.ascii.toLower(a_c) != std.ascii.toLower(b_c)) return false;
+    }
+    return true;
+}
+
+/// StaticStringMap, but accepts an equality function (`eql`).
+/// The `eql` function is only called to determine the equality
+/// of equal length strings. Any strings that are not equal length
+/// are never compared using the `eql` function.
+pub fn StaticStringMapWithEql(
+    comptime V: type,
+    comptime eql: fn (a: []const u8, b: []const u8) bool,
+) type {
+    return struct {
+        kvs: *const KVs = &empty_kvs,
+        len_indexes: [*]const u32 = &empty_len_indexes,
+        len_indexes_len: u32 = 0,
+        min_len: u32 = std.math.maxInt(u32),
+        max_len: u32 = 0,
+
+        pub const KV = struct {
+            key: []const u8,
+            value: V,
+        };
+
+        const Self = @This();
+        const KVs = struct {
+            keys: [*]const []const u8,
+            values: [*]const V,
+            len: u32,
+        };
+        const empty_kvs = KVs{
+            .keys = &empty_keys,
+            .values = &empty_vals,
+            .len = 0,
+        };
+        const empty_len_indexes = [0]u32{};
+        const empty_keys = [0][]const u8{};
+        const empty_vals = [0]V{};
+
+        /// Returns a map backed by static, comptime allocated memory.
+        ///
+        /// `kvs_list` must be either a list of `struct { []const u8, V }`
+        /// (key-value pair) tuples, or a list of `struct { []const u8 }`
+        /// (only keys) tuples if `V` is `void`.
+        pub inline fn initComptime(comptime kvs_list: anytype) Self {
+            comptime {
+                @setEvalBranchQuota(30 * kvs_list.len);
```

Comment:
@Vexu done.  This seems to be just big enough.  I've added another test at bottom with a kvs_list of length 1000.